### PR TITLE
[tune] Local Mode support

### PR DESCRIPTION
**Pull request by richardliaw**
Support Local Mode for Tune. This will enable much cleaner debugging.

### CODE REVIEW

File: python/ray/tune/ray_trial_executor.py [hunk 1]
```diff
@@ -229,6 +243,10 @@ def fetch_result(self, trial):
             raise ValueError("Trial was not running.")
         self._running.pop(trial_future[0])
         result = ray.get(trial_future[0])
+
+        # For local mode
+        if not type(result) is dict:
```
Comment (ericl) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
        if isinstance(result, _LocalWrapper):
```

**Review by ericl — APPROVED** [review]

File: python/ray/tune/examples/logging_example.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def _restore(self, checkpoint_path):
     parser.add_argument(
         "--smoke-test", action="store_true", help="Finish quickly for testing")
     args, _ = parser.parse_known_args()
-    ray.init()
+    ray.init(local_mode=True)
```
Comment (richardliaw) [inline python/ray/tune/examples/logging_example.py]:
```suggestion
    ray.init()
```

File: python/ray/tune/ray_trial_executor.py [hunk 3]
```diff
@@ -266,7 +284,13 @@ def _return_resources(self, resources):
 
     def _update_avail_resources(self, num_retries=5):
         for i in range(num_retries):
-            resources = ray.global_state.cluster_resources()
+            try:
+                resources = ray.global_state.cluster_resources()
+            except Exception:
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/ray_trial_executor.py]: Yeah, unfortunately that's what Ray throws right now...

ray/python/ray/experimental/state.py: L116

I'll open an issue for this.

File: python/ray/tune/ray_trial_executor.py [hunk 4]
```diff
@@ -266,7 +284,13 @@ def _return_resources(self, resources):
...
+                # TODO(rliaw): Remove this when local mode is fixed.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/ray_trial_executor.py]: Done

**Pull request merged**